torchmetrics/metric.py

```diff
@@ -411,6 +413,80 @@ def __setstate__(self, state: Dict[str, Any]) -> None:
         self.update: Callable = self._wrap_update(self.update)  # type: ignore
         self.compute: Callable = self._wrap_compute(self.compute)  # type: ignore
 
+    @property
+    def dtype(self) -> "torch.dtype":
+        """Return the dtype of the metric."""
+        return self._dtype
+
+    @dtype.setter
+    def dtype(self, new_dtype: "torch.dtype") -> None:
+        # necessary to avoid infinite recursion
+        raise RuntimeError("Cannot set the dtype explicitly. Please use module.to(new_dtype).")
+
+    @property
+    def device(self) -> "torch.device":
+        """Return the device of the metric."""
+        return self._device
+
+    def to(self, *args: Any, **kwargs: Any) -> "Metric":
+        """Moves and/or casts the parameters and buffers.
+
+        Works similar to nn.Module.to but also updates the metrics device and dtype properties
+        """
+        # there is diff nb vars in PT 1.5
+        out = torch._C._nn._parse_to(*args, **kwargs)
+        self._update_properties(device=out[0], dtype=out[1])
+        return super().to(*args, **kwargs)
+
+    def cuda(self, device: Optional[Union[torch.device, int]] = None) -> "Metric":
+        """Moves all model parameters and buffers to the GPU.
+
+        Arguments:
+            device: if specified, all parameters will be copied to that device
+        """
+        if device is None or isinstance(device, int):
+            device = torch.device("cuda", index=device)
+        self._update_properties(device=device)
+        return super().cuda(device=device)
+
+    def cpu(self) -> "Metric":
+        """Moves all model parameters and buffers to the CPU."""
+        self._update_properties(device=torch.device("cpu"))
+        return super().cpu()
+
+    def type(self, dst_type: Union[str, torch.dtype]) -> "Metric":
+        """Casts all parameters and buffers to :attr:`dst_type`.
+
+        Arguments:
+            dst_type (type or string): the desired type
+        """
+        self._update_properties(dtype=dst_type)
+        return super().type(dst_type=dst_type)
+
+    def float(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``float`` datatype."""
+        self._update_properties(dtype=torch.float)
+        return super().float()
+
+    def double(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``double`` datatype."""
+        self._update_properties(dtype=torch.double)
+        return super().double()
+
+    def half(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``half`` datatype."""
+        self._update_properties(dtype=torch.half)
+        return super().half()
+
+    def _update_properties(
```

Comment:
Given that we mostly just update one thing, will it be easier to just directly write `self._dtype=dtype` in those functions instead of calling this?